Restructure `sorted_schur`

Marius1311 opened the pull request:
I found a bug with using a pre-computed schur decomposition, with didn't have enough eigenvalues to check for block-splitting when the original decomp. was computed using `krylov`. I fixed that bug and decided to re-structure the `sorted_schur` functions a bit, so that `brandts` and `krylov` are more consistent in that...
- there are functions `sorted_krylov_schur` and `sorted_brandts_schur` which each return Q, R and the eigenvalues
- in both cases, we first compute the decomposition and then check for splitting blocks of complex conjugates, using the same utility function
- we execute the same checks on the computed schur decomposition, no matter whether we computed it using `brandts` or `krylov`

One thing I wanted your opinion on: in case we choose a number `m` that would split a block of complex conjugates, so far, we just raise an exception and the user has to start again. It may be more convenient to simply raise a warning and to use the schur decomposition of size `m+1`, which we have computed already. In case the user prefers using `m-1`, that's no problem either since we saved the decomposition, so it's only a matter of subsetting it. What do you think?

Marius1311 committed:
Introduce k=m+1

Marius1311 committed:
Create utility function for checking whether we cut compl. conj

Marius1311 committed:
REturn m+1 eigenvalues for krylov-schur

Marius1311 committed:
Deleta the top_eigenvalues function

Marius1311 committed:
Change the structure of the program

Marius1311 committed:
Move realness test

Marius1311 committed:
Fix bug with length of the eigenvalues

Marius1311 committed:
Compute m

Marius1311 committed:
Fix bug with subspace

Marius1311 commented:
I run this through the checks I run for the last two PRs and everything worked.

Marius1311 committed:
Leave the matrix sparse, only make it dense for schur if brandts

Marius1311 commented:
Hi @msmdev , would be great if you could have a look at this today. Thanks!

msmdev commented:
> Hi @msmdev , would be great if you could have a look at this today. Thanks!

@Marius1311 I will do my very best :)

msmdev commented on msmtools/util/sorted_schur.py:
```diff
@@ -23,111 +23,64 @@ def _initialize_matrix(M, P):
+    dummy4 = subspace_angles(dummy, Q)
+    test4 = np.allclose(dummy4, 0.0, atol=1e-6, rtol=1e-5)
+    if not test4:
+        raise ValueError(f"According to scipy.linalg.subspace_angles() `{method}` "
```
```suggestion
        raise ValueError(f"According to scipy.linalg.subspace_angles() `{method}`  didn't "
```

msmdev commented on msmtools/util/sorted_schur.py:
```diff
@@ -23,111 +23,64 @@ def _initialize_matrix(M, P):
+                      f"matrix with the sorted top eigenvalues on the diagonal), method = `{method}`")
+
+
+def sorted_krylov_schur(P, k, z='LM', tol=1e-16):
```
If you use k instead of m, please also update the docstring... :)

Marius1311 replied:
alright, will do!

msmdev commented on msmtools/util/sorted_schur.py:
```diff
@@ -227,82 +170,55 @@ def sorted_krylov_schur(P, m, z='LM', tol=1e-16):
+    if np.any(np.array(ap) > 1.0):
+        warnings.warn("Reordering of Schur matrix was inaccurate.")
+
+    # comptue eigenvalues
```
```suggestion
    # compute eigenvalues
```

msmdev commented on msmtools/util/sorted_schur.py:
```diff
@@ -23,111 +23,64 @@ def _initialize_matrix(M, P):
-
-def sorted_krylov_schur(P, m, z='LM', tol=1e-16):
+
+def _check_schur(P, Q, R, eigenvalues, method = ""):
```
Why do you use `method = ""` instead of `method`? A method should be passed in any case, right?

Marius1311 replied:
Since the `method` parameter is only used for printing in the `_check_schur` function, I made it optional so that I could use the function more flexibly, e.g. when I want to call it from outside `sorted_schur.py`.

Marius1311 replied:
I don't have a strong opinion here, if you want, I can also make it mandatory to pass the method string.

msmdev replied:
I guess it doesn't matter from point of safety. So do as you like :)

Marius1311 replied:
alright, I left it the way it had done it originally.

msmdev commented on msmtools/util/sorted_schur.py:
```diff
@@ -360,44 +276,29 @@ def sorted_schur(P, m, z='LM', method='brandts', tol_krylov=1e-16):
+    # check for splitting pairs of complex conjugates
+    if (m < n):
+        if _check_conj_split(m, eigenvalues):
+            raise ValueError(f'Clustering into {m} clusters will split conjugate eigenvalues. '
```
Yes, we could warn here instead of raising and further use Q[:, :m+1], R[:m+1, :m+1]...

Marius1311 replied:
I opted for not doing this yet - I think it would be cleaner to do this in a separate PR.

msmdev commented on msmtools/analysis/dense/gpcca.py:
```diff
@@ -1075,10 +1075,10 @@ def __init__(self, P, eta=None, z='LM', method='brandts'):
         if method not in ['brandts', 'krylov']:
             raise ValueError("You didn't give a valid method to determine the invariant subspace.")
           
-        if issparse(P) and method != 'krylov':
```
I just recognized that in `gpcca_coarsegrain` sparse matrices are not handled...

Marius1311 replied:
For now, I fixed this by just making it dense in case it's sparse. I added a todo and will open an issue for this, so we can do it properly it a separate PR.

msmdev commented on msmtools/util/sorted_schur.py:
```diff
@@ -23,111 +23,64 @@ def _initialize_matrix(M, P):
-        Options are:
-        'LM': the m eigenvalues with the largest magnitude are sorted up.
-        'LR': the m eigenvalues with the largest real part are sorted up.
+def _check_conj_split(m, eigenvalues):
```
I guess now we don't need `find_twoblocks` in `gpcca` anymore...

Marius1311 replied:
I removed the function and import `_check_conj_split` from `util`, to have a single source of truth. This is more clean, I think.

msmdev reviewed (commented):
Nice work! Just minor comments :)

Marius1311 committed:
Update msmtools/util/sorted_schur.py

Co-authored-by: Bernhard Reuter <43750470+msmdev@users.noreply.github.com>

Marius1311 committed:
Update msmtools/util/sorted_schur.py

Co-authored-by: Bernhard Reuter <43750470+msmdev@users.noreply.github.com>

Marius1311 committed:
Change doscring to mention k instead of m

Marius1311 committed:
Alternative check for splitting a pair of complex conjugates

Marius1311 committed:
Remove m argument from checking function

Marius1311 committed:
Only save m eigenvalues

Marius1311 committed:
Apply changes to GPCCA class

Marius1311 committed:
Switch to using the eigenvalue criterion everywhere

Marius1311 committed:
Remove the find_two block function

Marius1311 committed:
Remove comment about find_twoblocks (todo)

Marius1311 committed:
Increase m automaticaly

Marius1311 committed:
Revert "Increase m automaticaly"

This reverts commit 9fa5ead88b8d3963942938bccbf3a6d544678851.

Marius1311 committed:
Make eta optional in gpcca_coarsegrain

Marius1311 committed:
Make gpcca_coarsegrain work again

Marius1311 commented:
I think I addressed all of your comments. To make the implementation cleaner, I changed the check for splitting complex conjugates to work with m rather than m+1 eigenvalues. 

This will make it easier in the future if we decide that we just want to warn in case we split a block of complex conjugates and continue with a decomposition of size m+1 becasue in that case, we would just have m+1 eigenvalues, and the functions that assume that there is always one eigenvalue more than requested would fail, specifically if we use a pre-computed schur decomposition.

msmdev commented:
@Marius1311 thx for your nice work! :)

Marius1311 commented:
You're welcome! Glad that you liked it.
